### PR TITLE
Avoid duplicate shader created by customMaterial & pbrCustomMaterial

### PR DESCRIPTION
**Pull request by Starryi**
I define a new class extending `PBRCustomMaterial` and declare getter/setter for uniform value, as follows
```js
class TintColorifyPBRCustomMaterial extends BABYLON.PBRCustomMaterial {
    _tintColor = BABYLON.Color3.Black();
    set tintColor(v) {
        this._tintColor = v;
    }
    get tintColor() {
        return this._tintColor;
    }

    constructor(name, scene) {
        super(name, scene);

        this.unlit = true;

        this.AddUniform('tintColor', 'vec3', this.tintColor);
        this.Fragment_Before_FragColor(`
        finalColor.rgb = tintColor;
        `);

        this.onBindObservable.add(() => {
            const effect = this.getEffect();
            effect.setColor3('tintColor', this.tintColor);
        });
    }
}
```

In use, `TintColorifyPBRCustomMaterial` likes `PBRMaterial` and other builtin material types in babylon library, and is very convenient to use, like follows 
```js
const mat = new TintColorifyPBRCustomMaterial('mat', scene);
mat.tintColor = BABYLON.Color3.Red();
```
That is my main purpose to define new class.

but, the above code will insert duplicate shaders to `Effect.ShadersStore` in such case: 
```js
const mat1 = new TintColorifyPBRCustomMaterial('mat1', scene);
const mat2 = new TintColorifyPBRCustomMaterial('mat2', scene);
```
https://www.babylonjs-playground.com/#W90UFJ#1 can reproduce the issue.

In this commit, I hope to fix the problem.

When defining new class extending `PBRCustomMaterial` and `CustomMaterial`,  if you specify a shader name for your new material type,  by assigning a value to `this._createdShaderName`, as follows
```js
class TintColorifyPBRCustomMaterial extends BABYLON.PBRCustomMaterial {
    _tintColor = BABYLON.Color3.Black();
    set tintColor(v) {
        this._tintColor = v;
    }
    get tintColor() {
        return this._tintColor;
    }

    constructor(name, scene) {
        super(name, scene);

        this.unlit = true;

        this.AddUniform('tintColor', 'vec3', this.tintColor);
        this.Fragment_Before_FragColor(`
        finalColor.rgb = tintColor;
        `);

        this._createdShaderName = 'TingColorify';


        this.onBindObservable.add(() => {
            const effect = this.getEffect();
            effect.setColor3('tintColor', this.tintColor);
        });
    }
}
```
because `mat1` and `mat2` has same `_createdShaderName`, so no duplicate shaders in `Effect.ShadersStore`

If you do not assign a value to `_createdShaderName`, then the shader name resolving logic as before,  nothing is changed.

### CODE REVIEW

**Comment by Starryi** [comment]
because the undermentioned two snippets in `Builder` method are only be executed once, so they are moved into `constructor` method.

```js
const name: string = "custom_" + CustomMaterial.ShaderIndexer;
const name = this._createdShaderName;
```

```js
const fn_afterBind = this._afterBind.bind(this);
        this._afterBind = (m, e) => {
            if (!e) {
                return;
            }
            this.AttachAfterBind(m, e);
            try {
                fn_afterBind(m, e);
            } catch (e) {}
        };
```

File: packages/dev/materials/src/custom/customMaterial.ts [hunk 1]
```diff
@@ -185,19 +171,29 @@ export class CustomMaterial extends StandardMaterial {
...
-        this._createdShaderName = name;
-
         return name;
     }
 
     constructor(name: string, scene?: Scene) {
         super(name, scene);
         this.CustomParts = new ShaderSpecialParts();
         this.customShaderNameResolve = this.Builder;
+        const fn_afterBind = this._afterBind.bind(this);
```
Comment (RaananW) [inline packages/dev/materials/src/custom/customMaterial.ts:181]: there is no need to bind here. this._afterBind is already using an arrow function and thus attached to the context in which you have defined it.

Comment (Starryi) [reply packages/dev/materials/src/custom/customMaterial.ts:181]: _afterBind seems to be override like this
```js
protected _afterBind(mesh?: Mesh | undefined, effect?: Nullable<Effect>) {
        if (!effect) {
            return;
        }
        this.AttachAfterBind(mesh, effect);
        try {
            super._afterBind(mesh, effect);
        } catch (e) {}
}
```

File: packages/dev/materials/src/custom/customMaterial.ts [hunk 2]
```diff
@@ -198,6 +182,19 @@ export class CustomMaterial extends StandardMaterial {
...
+    }
+
+    protected _afterBind(mesh?: Mesh | undefined, effect?: Nullable<Effect>): void {
+        if (!effect) {
+            return;
+        }
+        this.AttachAfterBind(mesh, effect);
+        try {
+            super._afterBind(mesh, effect);
+        } catch (e) {}
```
Comment (RaananW) [inline packages/dev/materials/src/custom/customMaterial.ts:197]: might be a naive question, but usually when i see an empty catch i wonder - in what case is it ok for this to fail? shouldn't it be logged?

**Comment by bjsplat** [comment]
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14571/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

**Pull request merged**